File: backend/app/schemas/deudas_simples.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.db.models import DeudaSimpleEstado, Moneda
from app.schemas.cheques import ChequeRead
# El vuelto de un cheque "de más" se resuelve igual que en pasivos (§5): o se
# paga en efectivo, o el negocio queda debiendo. Se reusa el mismo tipo para que
# no aparezcan dos vocabularios para la misma decisión.
from app.schemas.pasivos import VueltoModo


class DeudaSimpleCreate(BaseModel):
    """Alta de una deuda libre de un cliente (sin cuotas ni cheque).

    Registrarla saca la plata de la caja (EGRESO en `moneda`, con fecha `fecha`).
    """

    cliente_id: UUID
    concepto: str = Field(min_length=1)
    monto: Decimal = Field(gt=0, max_digits=18, decimal_places=2)
    moneda: Moneda
    fecha: date | None = None
    observaciones: str | None = None


class DeudaSimpleUpdate(BaseModel):
    """Corrección de la carga de una deuda libre desde el panel.

    Campos opcionales (`exclude_unset`). `concepto`, `fecha` y `observaciones` se
    editan siempre. `monto`/`moneda` solo si está ABIERTA y sin cobros parciales
    (saldo == monto); el servicio recalcula el saldo y resincroniza el egreso."""

    concepto: str | None = Field(default=None, min_length=1)
    monto: Decimal | None = Field(default=None, gt=0, max_digits=18, decimal_places=2)
    moneda: Moneda | None = None
    fecha: date | None = None
    observaciones: str | None = None


class DeudaSimplePagoRequest(BaseModel):
    """Cobro de una deuda libre (total o parcial), en efectivo.

    `monto_cobrado` es lo que entra a caja, en `moneda_pago` (que puede diferir de
    la moneda de la deuda). `cotizacion` ($/USD) es obligatoria solo cuando
    `moneda_pago` ≠ moneda de la deuda; imputa cuánto del saldo queda saldado."""

    monto_cobrado: Decimal = Field(gt=0, max_digits=18, decimal_places=2)
    moneda_pago: Moneda
    cotizacion: Decimal | None = Field(default=None, gt=0, max_digits=18, decimal_places=4)
    # A cuánto entran al stock vendible los dólares cobrados (§Stock de dólares).
    # Obligatoria al cobrar en USD **una deuda que también es en USD**: ahí no hay
    # `cotizacion` de la que sacar el costo, y sin costo esos dólares no se pueden
    # vender. Cuando el cobro cruza monedas, la `cotizacion` de arriba ya sirve.
    cotizacion_stock: Decimal | None = Field(
        default=None, gt=0, max_digits=18, decimal_places=6
    )
    fecha_cobro: date | None = None


class DeudaSimpleCobroClienteCreate(BaseModel):
    """Cobro de un importe libre contra todas las deudas abiertas de un cliente.

    Es el cobro de la fila del cliente en "Otras deudas": el importe se imputa a
    las deudas de la más vieja a la más nueva, sin que el operador tenga que
    elegir a cuál va.

    `moneda_deuda` dice contra qué deudas se cobra: ARS y USD son cajas distintas
    y no se suman entre sí. `monto_cobrado` es lo que entra a caja, en
    `moneda_pago`, que puede diferir; en ese caso `cotizacion` ($/USD) imputa
    cuánto del saldo queda saldado."""

    cliente_id: UUID
    moneda_deuda: Moneda
    monto_cobrado: Decimal = Field(gt=0, max_digits=18, decimal_places=2)
    moneda_pago: Moneda
    cotizacion: Decimal | None = Field(default=None, gt=0, max_digits=18, decimal_places=4)
    # Costo de entrada al stock de los dólares cobrados; ver DeudaSimplePagoRequest.
    cotizacion_stock: Decimal | None = Field(
        default=None, gt=0, max_digits=18, decimal_places=6
    )
    fecha_cobro: date | None = None


class DeudaSimpleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    cliente_id: UUID
    concepto: str
    monto: Decimal
    saldo_pendiente: Decimal
    moneda: Moneda
    estado: DeudaSimpleEstado
    fecha: date
    fecha_cancelacion: date | None
    observaciones: str | None
    cotizacion_pago: Decimal | None
    created_at: datetime
    updated_at: datetime


class DeudaSimpleCobroClienteChequeCreate(BaseModel):
    """Cobro de TODAS las deudas abiertas de un cliente con un solo cheque.

    El cheque salda por su **valor neto** (`monto × (1 − %compra)`), imputado de
    la deuda más vieja a la más nueva. Un cheque que vale más que todo lo que el
    cliente debe es el caso normal —el cliente entrega el que tiene—, así que la
    diferencia queda a favor suyo y `vuelto_modo` decide qué se hace con ella
    (obligatorio solo si hay diferencia). Mismo mecanismo que el vuelto de un
    pasivo pagado con cheque de más (§5).

    Los cheques son siempre en pesos: cobrar deudas en USD exige `cotizacion`."""

    cliente_id: UUID
    moneda_deuda: Moneda
    nro_cheque_pago: str = Field(min_length=1, max_length=64)
    banco_pago: str | None = Field(default=None, max_length=120)
    monto_cheque: Decimal = Field(gt=0, max_digits=18, decimal_places=2)
    porcentaje_compra_cheque: Decimal = Field(ge=0, le=100, max_digits=7, decimal_places=4)
    fecha_emision: date | None = None
    fecha_pago: date | None = None
    cotizacion: Decimal | None = Field(default=None, gt=0, max_digits=18, decimal_places=4)
    vuelto_modo: VueltoModo | None = None
    fecha_cobro: date | None = None


class DeudaSimpleCobroClienteChequeResponse(BaseModel):
    """Resultado del cobro por cliente con cheque.

    `imputado` es cuánto bajó la deuda en total y `saldo_restante` lo que el
    cliente sigue debiendo, ambos en la moneda de las deudas. `vuelto_ars` va en
    **ARS** —el excedente de un cheque es plata en pesos— y es > 0 solo cuando el
    cheque cubrió todo; `vuelto_modo` dice qué se hizo con él.
    """

    deudas_afectadas: list[DeudaSimpleRead]
    cheque_ingresado: ChequeRead
    imputado: Decimal
    canceladas: int
    saldo_restante: Decimal
    vuelto_ars: Decimal
    vuelto_modo: VueltoModo | None = None


class DeudaSimpleCobroClienteResponse(BaseModel):
    """Resultado del cobro por cliente.

    `deudas_afectadas` son las deudas que recibieron parte del importe, en el
    orden en que se imputaron (la más vieja primero). `imputado` es cuánto bajó
    la deuda en total y `saldo_restante` lo que el cliente sigue debiendo, ambos
    en la moneda de las deudas.
    """

    deudas_afectadas: list[DeudaSimpleRead]
    imputado: Decimal
    canceladas: int
    saldo_restante: Decimal


class DeudaSimpleCobrarConChequeRequest(BaseModel):
    """Cobro de una deuda libre entregando un cheque en vez de efectivo.

    El cheque entra a cartera a nombre del cliente de la deuda y vale su
    `valor_neto = monto_cheque × (1 − porcentaje_compra_cheque / 100)`.

    Los cheques son SIEMPRE en pesos, así que cobrar con cheque una deuda en USD
    cruza monedas y exige `cotizacion` ($/USD) — igual que el cobro en efectivo.
    """

    nro_cheque_pago: str = Field(min_length=1, max_length=64)
    banco_pago: str | None = Field(default=None, max_length=120)
    monto_cheque: Decimal = Field(gt=0, max_digits=18, decimal_places=2)
    porcentaje_compra_cheque: Decimal = Field(ge=0, le=100, max_digits=7, decimal_places=4)
    fecha_emision: date | None = None
    fecha_pago: date | None = None
    # Obligatoria solo si la deuda es en USD (el cheque siempre entra en ARS).
    cotizacion: Decimal | None = Field(default=None, gt=0, max_digits=18, decimal_places=4)
    # Obligatorio solo si el cheque cubre de más: qué se hace con el excedente.
    # Mismo criterio que el cobro por cliente y que el vuelto de un pasivo (§5).
    vuelto_modo: VueltoModo | None = None
    fecha_cobro: date | None = None


class DeudaSimpleCobrarConChequeResponse(BaseModel):
    """Resultado del cobro con cheque.

    `diferencia` en la moneda de la deuda: > 0 el cheque valía más que el saldo y
    el negocio le queda debiendo esa diferencia al cliente; < 0 el cliente todavía
    debe el resto; 0 justo.

    `vuelto_ars` es esa misma diferencia positiva llevada a **pesos** (el
    excedente de un cheque es plata en pesos y en pesos se devuelve), y
    `vuelto_modo` dice qué se hizo con ella: pagarla o quedar debiéndola.
    """

    deuda: DeudaSimpleRead
    cheque_ingresado: ChequeRead
    diferencia: Decimal
    vuelto_ars: Decimal
    vuelto_modo: VueltoModo | None = None

File: frontend/src/api/deudas_simples.ts
import { apiFetch } from './client'
import type { Cheque, DeudaSimple, DeudaSimpleEstado, Moneda } from '../types'

export interface DeudaSimpleCreatePayload {
  cliente_id: string
  concepto: string
  monto: number
  moneda: Moneda
  fecha: string | null
  observaciones: string | null
}

export const getDeudasSimples = (estado?: DeudaSimpleEstado): Promise<DeudaSimple[]> =>
  apiFetch<DeudaSimple[]>(`/deudas-simples${estado ? `?estado=${estado}` : ''}`)

export const createDeudaSimple = (payload: DeudaSimpleCreatePayload): Promise<DeudaSimple> =>
  apiFetch<DeudaSimple>('/deudas-simples', { method: 'POST', body: JSON.stringify(payload) })

// Corrección de la carga. `monto`/`moneda` solo si está ABIERTA y sin cobros
// parciales (lo valida el backend).
export interface DeudaSimpleUpdatePayload {
  concepto?: string
  monto?: number
  moneda?: Moneda
  fecha?: string | null
  observaciones?: string | null
}

export const editarDeudaSimple = (id: string, payload: DeudaSimpleUpdatePayload): Promise<DeudaSimple> =>
  apiFetch<DeudaSimple>(`/deudas-simples/${id}`, { method: 'PATCH', body: JSON.stringify(payload) })

export interface CobrarDeudaSimplePayload {
  monto_cobrado: number
  moneda_pago: Moneda
  // Requerida solo si moneda_pago difiere de la moneda de la deuda ($/USD).
  cotizacion?: number | null
  fecha_cobro?: string | null
}

export const cobrarDeudaSimple = (id: string, payload: CobrarDeudaSimplePayload): Promise<DeudaSimple> =>
  apiFetch<DeudaSimple>(`/deudas-simples/${id}/cobrar`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

/**
 * Cobro de un importe libre contra TODAS las deudas abiertas de un cliente.
 *
 * Es el cobro de la fila del cliente: el importe se imputa de la deuda más
 * vieja a la más nueva, sin que el operador elija a cuál va. `moneda_deuda`
 * dice contra qué deudas se cobra (ARS y USD no se suman entre sí); el pago
 * puede venir en la otra moneda con su cotización, como en el cobro suelto.
 */
export interface CobrarDeudasClientePayload {
  cliente_id: string
  moneda_deuda: Moneda
  monto_cobrado: number
  moneda_pago: Moneda
  // Requerida solo si moneda_pago difiere de moneda_deuda ($/USD).
  cotizacion?: number | null
  fecha_cobro?: string | null
}

export interface CobroClienteResult {
  /** Las deudas que recibieron parte del importe, la más vieja primero. */
  deudas_afectadas: DeudaSimple[]
  /** Cuánto bajó la deuda en total, en la moneda de las deudas. */
  imputado: string
  canceladas: number
  saldo_restante: string
}

export const cobrarDeudasCliente = (payload: CobrarDeudasClientePayload): Promise<CobroClienteResult> =>
  apiFetch<CobroClienteResult>('/deudas-simples/cobrar-cliente', {
    method: 'POST',
    body: JSON.stringify(payload),
  })

/**
 * Cobro de TODAS las deudas abiertas de un cliente con un solo cheque.
 *
 * Salda por el valor neto del cheque, de la deuda más vieja a la más nueva. Si
 * cubre todo y sobra, `vuelto_modo` decide qué se hace con la diferencia (que
 * va en pesos, porque el cheque es un instrumento en pesos): pagarla en
 * efectivo —lo único que mueve la caja acá— o quedar debiéndola, que crea un
 * pasivo a favor del cliente. Mismo mecanismo que el vuelto de un pasivo.
 */
export type VueltoModo = 'SALDAR_EFECTIVO' | 'QUEDA_DEBIENDO'

export interface CobrarDeudasClienteConChequePayload {
  cliente_id: string
  moneda_deuda: Moneda
  nro_cheque_pago: string
  banco_pago?: string | null
  monto_cheque: number
  porcentaje_compra_cheque: number
  fecha_emision?: string | null
  fecha_pago?: string | null
  // Requerida solo si las deudas son en USD (el cheque siempre entra en pesos).
  cotizacion?: number | null
  // Obligatorio solo si el cheque cubre todo y sobra.
  vuelto_modo?: VueltoModo | null
  fecha_cobro?: string | null
}

export interface CobroClienteChequeResult {
  deudas_afectadas: DeudaSimple[]
  cheque_ingresado: Cheque
  imputado: string
  canceladas: number
  saldo_restante: string
  /** En ARS: > 0 el cheque cubrió todo y el negocio le queda debiendo esto. */
  diferencia: string
  vuelto_modo: VueltoModo | null
}

export const cobrarDeudasClienteConCheque = (
  payload: CobrarDeudasClienteConChequePayload,
): Promise<CobroClienteChequeResult> =>
  apiFetch<CobroClienteChequeResult>('/deudas-simples/cobrar-cliente-con-cheque', {
    method: 'POST',
    body: JSON.stringify(payload),
  })

/**
 * Cobro de una deuda libre entregando un cheque en vez de efectivo.
 *
 * El cheque entra a cartera a nombre del cliente y salda por su valor neto
 * (`monto × (1 − %compra)`), no por su nominal. No mueve la caja: la plata se
 * reconoce recién cuando ese cheque se venda o se cobre.
 */
export interface CobrarDeudaSimpleConChequePayload {
  nro_cheque_pago: string
  banco_pago?: string | null
  monto_cheque: number
  porcentaje_compra_cheque: number
  fecha_emision?: string | null
  fecha_pago?: string | null
  // Requerida solo si la deuda es en USD (el cheque siempre entra en pesos).
  cotizacion?: number | null
  fecha_cobro?: string | null
}

export interface CobrarConChequeResult {
  deuda: DeudaSimple
  cheque_ingresado: Cheque
  /** En la moneda de la deuda: > 0 el negocio le queda debiendo al cliente. */
  diferencia: string
}

export const cobrarDeudaSimpleConCheque = (
  id: string,
  payload: CobrarDeudaSimpleConChequePayload,
): Promise<CobrarConChequeResult> =>
  apiFetch<CobrarConChequeResult>(`/deudas-simples/${id}/cobrar-con-cheque`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
